### PR TITLE
[#130267877] Automate creation of SSH tunnels

### PR DESCRIPTION
**Pull request by saliceti**
## What

When providing support for the platform, we often need to create SSH tunnels via the Concourse host to access internal IPs, internal ELBs, or external DNS resolving to internal IPs.

This extends the `ssh.sh` script to allow creation and deletion of SSH tunnels.

## How to review
Use case: get Elastic search cluster status.

* Export usual environment variables: DEPLOY_ENV, AWS keys, etc
* Create tunnel to known IP:

```
$ make dev tunnel TUNNEL=9200:10.0.16.10:9200
download: s3://master-state/concourse_id_rsa to ../../../../../../../tmp/concourse_id_rsa.9975
Creating tunnel at socket /Users/colin/ssh_tunnel_socket.ctl to 9200:10.0.16.10:9200
Warning: Permanently added '52.49.176.99' (ECDSA) to the list of known hosts.
Unauthorized use is strictly prohibited. All access and activity
is subject to logging and monitoring.
```

* Check ES status via tunnel:

```
$ curl localhost:9200/_cluster/health?pretty=true
{
  "cluster_name" : "logsearch",
  "status" : "green",
  "timed_out" : false,
  "number_of_nodes" : 5,
  "number_of_data_nodes" : 3,
  "active_primary_shards" : 151,
  "active_shards" : 302,
  "relocating_shards" : 0,
  "initializing_shards" : 0,
  "unassigned_shards" : 0,
  "delayed_unassigned_shards" : 0,
  "number_of_pending_tasks" : 0,
  "number_of_in_flight_fetch" : 0,
  "task_max_waiting_in_queue_millis" : 0,
  "active_shards_percent_as_number" : 100.0
}
```

* Stop the tunnel:

```
$ make dev stop-tunnel 
download: s3://master-state/concourse_id_rsa to ../../../../../../../tmp/concourse_id_rsa.26717
Stopping tunnel at socket /Users/colin/ssh_tunnel_socket.ctl
Exit request sent.
```

* Confirm the tunnel is down:

```
$ curl localhost:9200/_cluster/health?pretty=true
curl: (7) Failed to connect to localhost port 9200: Connection refused
```

* You can test it works with DNS and ELB too, such as:

```
$ make ci tunnel TUNNEL=9200:internal-master-logsearch-es-master-937682804.eu-west-1.elb.amazonaws.com:9200
```

## Who can review

Anyone but myself.

### CODE REVIEW

File: concourse/scripts/ssh.sh [hunk 1]
```diff
@@ -1,16 +1,45 @@
 #!/bin/bash
 
 set -euo pipefail
-key=/tmp/concourse_id_rsa.$RANDOM
-trap 'rm -f $key' EXIT
+TUNNEL=${1:-}
+SOCKET=~/ssh_tunnel_socket.ctl
```
Comment (alext) [inline concourse/scripts/ssh.sh]: Can this socket be put somewhere other then our homedir.  Maybe under `$TMPDIR`, or somewhere within the paas-cf directory tree.

Comment (saliceti) [inline concourse/scripts/ssh.sh]: As discussed, I've moved the socket to the default SSH place within .ssh

File: concourse/scripts/ssh.sh [hunk 2]
```diff
@@ -1,16 +1,51 @@
...
-ruby -ryaml -e 'puts "\Sudo password is " + YAML.load(STDIN)["secrets"]["concourse_vcap_password_orig"]' 
-echo
+  aws s3 cp "s3://${DEPLOY_ENV}-state/concourse_id_rsa" $key && chmod 400 $key
+}
 
-ssh -i $key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null vcap@"$CONCOURSE_IP"
+ssh_concourse() {
+  echo
+  aws s3 cp "s3://${DEPLOY_ENV}-state/generated-concourse-secrets.yml" - | \
+  ruby -ryaml -e 'puts "\Sudo password is " + YAML.load(STDIN)["secrets"]["concourse_vcap_password_orig"]'
```
Comment (mtekel) [inline concourse/scripts/ssh.sh]: There's no need for \ before S in '\Sudo password is'. Not sure why I had it there, but given that you have a beautifying commit in this PR, we can also beautify and unconfuse this.

**Comment by mtekel** [comment]
`make dev tunnel 8080:10.0.49.101:8080` does ssh to concourse instead...
Maybe do some checking in makefile before the script is called?

**Comment by mtekel** [comment]
btw. I'd squash in the blank line & backslash commits into the main commits doing changes in respective files.

**Comment by saliceti** [comment]
Done

**Comment by jimconner** [comment]
Is the README.md getting an update for this new feature?

**Comment by mtekel** [comment]
when you type make it lists all targets with help, that should be enough

**Comment by jimconner** [comment]
Yeah... tying make is good... but I thought that the examples in the 'How to Review' section of this PR would be good to go in the README.

**Comment by saliceti** [comment]
@jimconner you're right, it should be documented. But there are lots of other things not documented in this README, and we want to keep it minimal. I think in another story we should create a doc/operator.md, and document all the tips and tricks.

For the moment I've added a small section as suggested by @mtekel.

**Comment by mtekel** [comment]
thanks, merging

**Pull request merged**